Fix AI spamming harvester resource searches

reaperrr opened the pull request:
Fixes #14467.

reaperrr committed:
Move up harv.IsEmpty check in HackyAI

This is a fairly cheap check, so we should perform it before the activity checks

reaperrr committed:
Make the AI only consider harvesters 'idle' if last resource search failed

Otherwise the AI would consider the harvester 'idle' in too many situations.
This way, the AI now only uses its own resource search algorithm if the next resource patch is too far away for the FindResources activity to find it.

MustaphaTR reviewed (approved):
Looks good to me. 👍

reaperrr committed:
Make AI skip harvester orders if map has no resources

reaperrr commented:
Updated, now also covers the case @airetaM pointed out in #14467.
The AI now checks if there's even a ResourceLayer - and if yes, whether there are any cells with resources - before giving orders to harvesters.

The difference on maps without resources is outright massive. I'd actually like to see this in the playtest.

Smittytron reviewed (approved):
Tested this vs current release on map without resources. Easy to see the performance improvement. 👍

MustaphaTR reviewed (changes requested):
Something is off with 3rd commit. AI is no longer doing anything with harvs.

I have something like that in a testmap, also a 3x3 patch at other side of the map, after here finishes AI no longer searches for other one. 
![resim](https://user-images.githubusercontent.com/7933210/33710568-0ac3e170-db52-11e7-9b4b-a86dbc3f716d.png)

Similarly ones it built from WFac don't go to harvesting there, as FindResources can't find the other ore on the map.

pchote commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -251,6 +253,8 @@ public ResourceType Harvest(CPos cell)
 			if (c.Type == null)
 				return null;
 
+			--resCells;
```
Shouldn't this be inside the `if` below?

MustaphaTR replied:
Looks like indeed that's the reason. My testcase works fine with this moved there.

reaperrr committed:
Fix AI crashing when trying to make parachuted harvester search map for resources while mid-air

reaperrr commented:
Updated.

Also added a fix for https://github.com/OpenRA/OpenRA/issues/14467#issuecomment-349967217 (at some point the `IsInAir` boolean might come in handy in other places, too).

pchote commented on OpenRA.Mods.Common/Traits/World/ResourceLayer.cs:
```diff
@@ -288,6 +293,11 @@ public int GetMaxResourceDensity(CPos cell)
 			return Content[cell].Type.Info.MaxDensity;
 		}
 
+		public bool IsResourceLayerEmpty()
```
Nit: this would be better as a property than a method.

pchote reviewed (dismissed):
Works as advertised :+1:

Just one minor code nit:

reaperrr commented:
Updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9f74c1a4b45270f22e43aa896ce9415033cf4871...05866958643bd1f25f93142aa4b8eba3ea95b396)
